Support common rl interface

findmyway opened the pull request:
Some optional functions are note supported yet:

- [x] `state`
- [ ] `set_state!`

I removed the concept of RLBase.observation in the latest PR, so `state` is kind of vague now and I can't find an exact match of state in CommonRLInterface.

`AbstractMarkovEnv` and `AbstractZeroSumEnv` are not supported yet. It's not very difficult. I can add another two structs named `CommonRLMarkovEnv` and `CommonRLZeroSumEnv` and automatically switch when `Base.convert(::Type{CRL.AbstractEnv}, env::AbstractEnv)`. But considering that the related APIs are not clear now. I will add them later.

findmyway commented:
It seems that the lower bound of Julia version v1.4 in COmmonRLInterface breaks the CI here


https://github.com/JuliaReinforcementLearning/CommonRLInterface.jl/blob/d7e482d75bda557651cb87e329ea85cbf3cdbe1c/Project.toml#L11

zsunberg commented:
> It seems that the lower bound of Julia version v1.4 in COmmonRLInterface breaks the CI here

Shoot, I don't remember why I have that - maybe it is possible to work around. Are you trying to support Julia 1.0?

zsunberg commented on src/CommonRLInterface.jl:
```diff
@@ -0,0 +1,61 @@
+
+Base.convert(::Type{CRL.AbstractEnv}, env::AbstractEnv) = convert(CommonRLEnv, env)
+
+function Base.convert(::Type{CommonRLEnv}, env::AbstractEnv)
+    new_env = CommonRLEnv(env)
+    if ActionStyle(env) === FullActionSet()
+        @eval CRL.@provide CRL.valid_actions(x::typeof($new_env)) = get_legal_actions(x.env)
+        @eval CRL.@provide CRL.valid_action_mask(env::typeof($new_env)) = get_legal_actions_mask(env.env)
+    end
+    new_env
+end
```
This is a case where I think it's best to manually implement `provided`

```suggestion
Base.convert(::Type{CommonRLEnv}, env::AbstractEnv) = CommonRLEnv(env)

CRL.provided(::typeof(CRL.valid_actions), env::CommonRLEnv) = ActionStyle(env.env) === FullActionSet()
CRL.valid_actions(x::typeof($new_env)) = get_legal_actions(x.env)

CRL.provided(::typeof(CRL.valid_action_mask), env::CommonRLEnv) = ActionStyle(env.env) === FullActionSet()
CRL.valid_action_mask(env::typeof($new_env)) = get_legal_actions_mask(env.env)
```

findmyway replied:
Good suggestion!

findmyway commented:
> Looks great! I made a suggestion to avoid `@eval`, and I think it should use `AbstractMarkovEnv` and `AbstractZeroSumEnv`.

Added.

> Regarding `AbstractMarkovEnv`/`AbstractZeroSumEnv`, I think there should be a switch in `convert`. If your env has `get_num_players(env) == 1`, it should be a `AbstractMarkovEnv`. If it has `get_num_players(env) == 2 && UtilityStyle(env) == ZERO_SUM`, then it should be an `AbstractZeroSumEnv`, I think.

Updated.

> Also, if you feel like the `AbstractMarkovEnv` vs `AbstractZeroSumEnv` issue is really confusing, we should discuss it on the CommonRLEnvironment repo - we don't want to go down the wrong road.

It's not confusing. It's just that I can't find any functions tied to them yet. 😄 

> I would not expect it to be easy to support `CRL.state` and `CRL.setstate!` for a general RL framework, so I don't think you should worry about implementing those. If you have `InformationStyle(env) == PERFECT_INFORMATION`, then it might make sense to have `state` return the same thing as observe.

Done.

> Shoot, I don't remember why I have that - maybe it is possible to work around. Are you trying to support Julia 1.0?

No, since I used the functional object related methods in RLBase, the lower bound of Julia is restricted to v1.3. However, I think CRL doesn't introduce any specific grammar after v1.4, it would be better to drop that constraint. https://github.com/JuliaReinforcementLearning/CommonRLInterface.jl/pull/31

zsunberg commented on src/CommonRLInterface.jl:
```diff
@@ -10,30 +10,47 @@ struct CommonRLEnv{T<:AbstractEnv} <: CRL.AbstractEnv
+CRL.@provide CRL.reset!(env::CommonRLEnvs) = reset!(env.env)
+CRL.@provide CRL.actions(env::CommonRLEnvs) = get_actions(env.env)
+CRL.@provide CRL.observe(env::CommonRLEnvs) = get_state(env.env)
+CRL.@provide CRL.state(env::CommonRLEnvs) = get_state(env.env)
```
```suggestion
CRL.state(env::CommonRLEnvs) = get_state(env.env)
CRL.provided(::typeof(CRL.state), env::CommoRLEnvs) = InformationStyle(env.env) === PERFECT_INFORMATION
```

findmyway replied:
Thanks!

zsunberg commented on src/CommonRLInterface.jl:
```diff
@@ -47,7 +64,7 @@ end
 Base.convert(::Type{RLBaseEnv}, env::CRL.AbstractEnv) = RLBaseEnv(env, 0.f0)  # can not determine reward ahead. Assume `Float32`.
 
-get_state(env::RLBaseEnv) = CRL.observe(env.env)
+get_state(env::RLBaseEnv) = InformationStyle(env) === PERFECT_INFORMATION ? CRL.state(env.env) : CRL.observe(env.env)
```
Hmm... I dont think this is the best thing to do. I think it should always be `observe`. All of your algorithms use `get_state` to get information to make a decision on, right?

If someone implemented a partially observable environment in `CommonRLInterface`, they would expect decision to be made only on the observation.
```suggestion
get_state(env::RLBaseEnv) = CRL.observe(env.env)
```

findmyway replied:
> If someone implemented a partially observable environment in CommonRLInterface, they would expect decision to be made only on the observation.

You are correct here. Originally I thought only `CRL.state` is provided for POMDP environments. After checking the doc again, I find that 'CRL.observe` is always required. Thank you for your suggestion here!

> All of your algorithms use get_state to get information to make a decision on, right?

That's right. But I'm also planning to add more algorithms that rely on some other kinds of information.

zsunberg commented on src/CommonRLInterface.jl:
```diff
@@ -58,4 +75,7 @@ reset!(env::RLBaseEnv) = CRL.reset!(env.env)
+ActionStyle(env::RLBaseEnv) = CRL.provided(CRL.valid_actions, env.env) ? FullActionSet() : MinimalActionSet()
+UtilityStyle(env::RLBaseEnv{<:CRL.AbstractZeroSumEnv}) = ZERO_SUM
+InformationStyle(env::RLBaseEnv) = IMPERFECT_INFORMATION
+InformationStyle(env::RLBaseEnv{<:CRL.AbstractMarkovEnv}) = PERFECT_INFORMATION
```
As of now, CommonRLInterface doesn't have the traits necessary to query whether `AbstractMarkovEnv` is an MDP or POMDP - it seems like this should be `IMPERFECT_INFORMATION`.

findmyway replied:
I see. I agree that `IMPERFECT_INFORMATION` is a better default value.

zsunberg commented on src/CommonRLInterface.jl:
```diff
@@ -58,4 +75,7 @@ reset!(env::RLBaseEnv) = CRL.reset!(env.env)
-ActionStyle(env::RLBaseEnv) = CRL.provided(CRL.valid_actions, env.env) ? FullActionSet() : MinimalActionSet()
\ No newline at end of file
+ActionStyle(env::RLBaseEnv) = CRL.provided(CRL.valid_actions, env.env) ? FullActionSet() : MinimalActionSet()
+UtilityStyle(env::RLBaseEnv{<:CRL.AbstractZeroSumEnv}) = ZERO_SUM
```
Should there also be something for
`UtilityStyle(env::RLBaseEnv{<:CRL.AbstractMarkovEnv})`?

findmyway replied:
The default one is `GENERAL_SUM`. I'll add it here to avoid confusion. 👍
